diagnostics/include/performance_handler.h

```diff
@@ -0,0 +1,201 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0
+// License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+#include <map>
+#include <memory>
+#include <mutex>
+#include <ostream>
+#include <stdexcept>
+
+#include <hdr/hdr_interval_recorder.h>
+
+#include "assertUtils.hpp"
+
+namespace concord::diagnostics {
+
+// A recorder is a thread-safe type that should always be created in a shared pointer to ensure
+// proper destruction. The recorder is used to add values to the histogram in the recording thread.
+// Interval histograms can be extracted in other threads.
+struct Recorder {
+  // These values come directly from hdr histogram.
+  // https://github.com/HdrHistogram/HdrHistogram_c/blob/master/src/hdr_interval_recorder.h#L28-L32
+  Recorder(int64_t lowest_trackable_value, int64_t highest_trackable_value, int significant_figures) {
+    ConcordAssert(lowest_trackable_value > 0);
+    hdr_interval_recorder_init_all(&recorder, lowest_trackable_value, highest_trackable_value, significant_figures);
+  }
+
+  ~Recorder() { hdr_interval_recorder_destroy(&recorder); }
+  Recorder(const Recorder&) = delete;
+  Recorder& operator=(const Recorder&) = delete;
+
+  void record(int64_t val) { hdr_interval_recorder_record_value(&(recorder), val); }
+
+  hdr_interval_recorder recorder;
+};
+
+struct Histogram {
+  Histogram(const std::shared_ptr<Recorder>& recorder)
+      : recorder(recorder), start(std::chrono::system_clock::now()), snapshot_start(start), snapshot_end(start) {
+    snapshot = hdr_interval_recorder_sample_and_recycle(&(recorder->recorder), snapshot);
+    hdr_init(
+        snapshot->lowest_trackable_value, snapshot->highest_trackable_value, snapshot->significant_figures, &history);
+  }
+
+  ~Histogram() {
+    hdr_close(snapshot);
+    hdr_close(history);
+    snapshot = nullptr;
+    history = nullptr;
+  }
+
+  Histogram(const Histogram&) = delete;
+  Histogram& operator=(const Histogram&) = delete;
+
+  void takeSnapshot() {
+    snapshot_start = snapshot_end;
+    snapshot_end = std::chrono::system_clock::now();
+    // Add the previous snapshot to the history
+    hdr_add(history, snapshot);
```

Comment:
Since the histograms are the exact same size/shape this should never return discarded entries (return value > 0). I don't think it warrants an assert, as we don't want to crash. But it may be worth logging.

Reply:
Logs added.